### ISSUE TITLE
Db2 z/OS: Unable to find License File on Docker

### ISSUE THREAD

**Issue by lweir-hfcpartners** [ibmdb/python-ibmdb#933]
I am trying to run this as a Docker container under GCP Dataproc (performing a JDBC connection to z/OS). 
I have a valid 11.5 license file, and can run this successfully using Docker in Google cloud under CloudRun. Thus, I have followed all the steps successfully and am able to connect and access Db2. However, under the container in Dataproc, something is not able to find the license file correctly and I am getting the standard error message:

[IBM][CLI Driver] SQL1598N  An attempt to connect to the database server failed because of a licensing problem.  SQLSTATE=42968 SQLCODE=-1598

- Operating System Name: Debian 12, and also used Ubuntu 22.04 base images

- db2level output from clidriver if in use:
DB21085I  This instance or install (instance name, where applicable: "*") uses 
"64" bits and DB2 code release "SQL11059" with level identifier "060A010F".
Informational tokens are "DB2 v11.5.9.0", "special_36648", 
"DYN2311160803AMD64_36648", and Fix Pack "0".
Product is installed at 
"/opt/miniconda3/lib/python3.10/site-packages/clidriver".

- Target Db2 Server Version: DSN12015
- Python Version:: Python 3.10.14
- ibm_db version: ibm_db                         3.2.3
- For non-Windows, output of below commands:
  uname: Linux
  uname -m: x86_64
- Value of below environment variables if set:
  IBM_DB_HOME: /usr/local/lib/python3.10/site-packages/clidriver
  PATH: /usr/lib/spark/bin:/usr/lib/jvm/temurin-11-jdk-amd64/bin:/usr/local/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
  LIB/LD_LIBRARY_PATH/DYLD_LIBRARY_PATH: /usr/local/lib/python3.10/site-packages/clidriver/lib:

License file is based on version 11.5

/opt/miniconda3/lib/python3.10/site-packages/clidriver/license

Permissions on license file:
-rw-r--r-- 1 root root   917 May  2 13:45 db2consv_ee.lic

From the running code we are able to view the first file of the license file:

head -1 /opt/miniconda3/lib/python3.10/site-packages/clidriver/license/db2consv_ee.lic
[LicenseCertificate]

What I am looking to understand is how the code is validating the license file so I can work out what is different with running the Docker container under the Google Dataproc environment where this is working elsewhere without any issue. The same license file works without any issue elsewhere and I have validated the environment variables, and permissions.

I am looking to see when we are doing a connect, what is validated in that step since this is not part of the python code that I can see here?

Is it using any external commands or the cldriver programs as part of this process that I can look at to see what is missing as I can confirm the license file and permission is in place where the documentation is specifying and known to work elsewhere.

**Comment by bimalkjha** [comment]
@lweir-hfcpartners To know details about license requirement and how it works, you can check this ibm_db documentation: https://github.com/ibmdb/python-ibmdb?tab=readme-ov-file#for-zos-and-iseries-connectivity-and-sql1598n-error

In the shared info, we can see `IBM_DB_HOME: /usr/local/lib/python3.10/site-packages/clidriver`. Have you set IBM_DB_HOME before running `pip install ibm_db`? Because as per ibm_db documentation, if IBM_DB_HOME environment variable is set, ibm_db will not download clidriver under site_packages directory.

So, you have two versions of clidriver in your system: /usr/local/lib/python3.10/site-packages/clidriver and /opt/miniconda3/lib/python3.10/site-packages/clidriver but you have copied license in /opt/miniconda3/lib/python3.10/site-packages/clidriver only. May be at runtime, clidriver from /usr/local/lib/python3.10/site-packages/clidriver is getting used, so you need to copy license file under /usr/local/lib/python3.10/site-packages/clidriver/license directory.

To confirm that which clidriver is in effect and license is getting read or not, you can run the db2cli validate command too as below:
```
db2cli validate -database "dbname:hostname:port" -userid dbuser -passwd dbpass -connect -displaylic
```
Update database connection info in above command and share complete output here. Thanks.

**Comment by lweir-hfcpartners** [comment]
Sorry about some of the incorrect information there. I was testing out using native Python install and Conda so copied in some incorrect paths.

To answer your questions above.

IBM_DB_HOME: /opt/miniconda3/lib/python3.10/site-packages/clidriver

This was set after the pip install ibm_db

/opt/miniconda3/lib/python3.10/site-packages/clidriver is the only location on the machine (this was my error copying the output from the different test run)

Here is the output from the db2cli command - at the bottom (I just renamed the IP address below in the output). This shows it is unable to find the license file.

The following is the output from the running code showing the details of the license file and location:

List license directory: /opt/miniconda3/lib/python3.10/site-packages/clidriver/license
total 40
drwxr-xr-x 1 root root  4096 May 16 12:45 .
drwxr-xr-x 1 root root  4096 May 16 12:45 ..
drwxr-xr-x 1 root root  4096 May 16 12:45 UNIX
-rw-r--r-- 1 root root   917 May 15 13:10 db2consv_ee.lic
-rwxrwxr-x 1 root root  3442 Mar 13 14:28 odbc_REDIST.txt
-rwxrwxr-x 1 root root 23445 Mar 13 14:28 odbc_notices.txt

Output of the running code is able to open and print the first line of the license file shown below:

License output: /opt/miniconda3/lib/python3.10/site-packages/clidriver/license/db2consv_ee.lic
[LicenseCertificate]


Output of the db2cli command.

===============================================================================
Client information for the current copy:
===============================================================================

Client Package Type       : IBM Data Server Driver For ODBC and CLI
Client Version (level/bit): DB2 v11.5.9.0 (special_36648/64-bit)
Client Platform           : Linux/X8664
Install/Instance Path     : /opt/miniconda3/lib/python3.10/site-packages/clidriver
DB2DSDRIVER_CFG_PATH value: <not-set>
db2dsdriver.cfg Path      : /opt/miniconda3/lib/python3.10/site-packages/clidriver/cfg/db2dsdriver.cfg
DB2CLIINIPATH value       : <not-set>
db2cli.ini Path           : /opt/miniconda3/lib/python3.10/site-packages/clidriver/cfg/db2cli.ini
db2diag.log Path          : /opt/miniconda3/lib/python3.10/site-packages/clidriver/db2dump/db2diag.log

===============================================================================
db2dsdriver.cfg schema validation for the entire file:
===============================================================================

Note: The validation utility could not find the configuration file 
db2dsdriver.cfg. The file is searched at 
"/opt/miniconda3/lib/python3.10/site-packages/clidriver/cfg/db2dsdriver.cfg".

===============================================================================
db2cli.ini validation for database "DB2Q:X.X.X.X:452":
===============================================================================

Note: The validation utility could not find the configuration file db2cli.ini. 
The file is searched at 
"/opt/miniconda3/lib/python3.10/site-packages/clidriver/cfg/db2cli.ini".

===============================================================================
db2dsdriver.cfg validation for database "DB2Q:X.X.X.X:452":
===============================================================================

Note: The validation utility could not find the configuration file 
db2dsdriver.cfg. The file is searched at 
"/opt/miniconda3/lib/python3.10/site-packages/clidriver/cfg/db2dsdriver.cfg".

===============================================================================
Connection attempt for database "DB2Q:X.X.X.X:452":
===============================================================================

[FAILED]: [IBM][CLI Driver] SQL1598N  An attempt to connect to the database server failed because of a licensing problem.  SQLSTATE=42968

===============================================================================

Valid License Found            : No
License location               : N/A
Server routine type attempted  : SP and UDF
DB2 connect license version    : N/A

===============================================================================
Error: The validation operation failed.
===============================================================================

===============================================================================
The validation is completed.
===============================================================================




This is where I am trying to understand how it is looking where the license file is located as I have followed the instructions in the link you provided and running the debug output from the python code clearly shows the license file is installed and readable. 

The db2cli command was executed from the same python code using subprocess.run() to execute in the same context as the executing python code.

**Comment by bimalkjha** [comment]
@lweir-hfcpartners ibm_db driver will try to read db2con*.lic file from location `/opt/miniconda3/lib/python3.10/site-packages/clidriver/license/` as license file. So the user who is running application should have read permission for this license file.
Also, after reading the license file, clidriver creates some cache file under `site-packages/clidriver/license/` and `site-packages/clidriver/cfgcache` directory. So, the user who runs python program should have write permission for clidriver/license and clidriver/cfgcache directory. Please share output of below commands to check permissions:

```
ls -ld /opt/miniconda3/lib/python3.10/site-packages/clidriver/license
ld -ld /opt/miniconda3/lib/python3.10/site-packages/clidriver/cfgcache
echo " " > /opt/miniconda3/lib/python3.10/site-packages/clidriver/license/a.txt
echo " " > /opt/miniconda3/lib/python3.10/site-packages/clidriver/cfgcache/b.txt
ls -l /opt/miniconda3/lib/python3.10/site-packages/clidriver/license
ls -l /opt/miniconda3/lib/python3.10/site-packages/clidriver/cfgcache
```
Thanks.

**Comment by lweir-hfcpartners** [comment]
I had tried earlier to make the license directory permissions 777 but that didn't work.

I updated the clidriver/cfgcache directory so the running user of the python code has write access and that has solved the issue. It is now able to recognize the license file and connect to the DB2 server and execute the remainder of the code.

That was the missing link. A big thank you on helping to resolve this issue. 

It may be useful for others to update the list of requirements in the link above with this additional piece of information.
